### DISCUSSION TITLE
Webhook push event - interrupt - validation - carry on push

### DISCUSSION THREAD

**Discussion by Molama** [community/community#24410] [Apps, API and Webhooks]
<p>Heja!</p>
<p>My scenario (as i would like it to be) is: </p>
<ul>
<li>webhook sends push event to url</li>
<li>server at specified url receives payload</li>
<li>push to repository is on hold</li>
<li>validations at server are executed and if they are NOT OK</li>
<li>do not execute the push to the repository/ status of files should be as before the try to push</li>
</ul>
<p>Is this scenario somehow possible or is it better to use a pre-receive hook?</p>
<p>Are webhooks serving as a kind of “information-service” after the event has happened and there is no way to prevent the event-action to happen if validations or anything else went wrong?</p>
<p>Would it be an idea to ammend the commit if the validations are NOT OK?</p>
<p>Thank you!</p>

**Comment by MadhuKulkarni** [comment]
<p>even I am having similar use case. Please let me know if you find a solution for this.</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>On <a href="http://github.com" rel="nofollow noopener">github.com</a>, there isn’t support for the scenario you’re describing. On premise installations of GitHub Enterprise can use pre-receive hooks for this kind of thing, I believe.</p>
<p>Yes, webhooks serve as notifications of events that have already occurred. There is no webhook analog of “event will happen if you return true or will be canceled if you return false”.</p>
<p>The standard way to do the kind of thing you’re describing is:</p>
<ol>
<li>A push occurs</li>
<li>Webhook notifies a service</li>
<li>Service checks the head commit</li>
<li>Service reports a status on the head commit</li>
</ol>
<p>That status record can be used in other logic for things like, “Don’t allow merges of PRs with failed required status checks”.</p>
<p>One reason why the current implementation is better than your proposed implementation would be:</p>
<ol>
<li>I make some changes to the project on my desktop computer and run local tests</li>
<li>Local tests fail</li>
<li>I have to leave the office right now to catch a plane, go on vacation, take care of a family emergency, etc</li>
<li>I push the local changes to the central repository so that I can pull them down on my laptop later</li>
<li>The central repository blocks my push because tests are failing</li>
</ol>
<p>Now I’m at the airport, at home at a lull in the family emergency, or a coworker tries to get my latest changes … but my latest changes aren’t on the central repository because the push was blocked. So no progress can be made until I can get back to my desktop at the office or someone can recreate my work.</p>
<p>But if the push is allowed, then no work stoppage occurs.</p>
<p>GitHub supports a system of <a href="https://help.github.com/en/articles/about-protected-branches" rel="nofollow noopener">protected branches</a> for modeling the kinds of workflows it sounds like you’re interested in. Branch protections are about preventing merges to specific branches, such as your repository’s default branch, so that the protected branch will always be clean and buildable, but people are free to push whatever they need to outside of the protected branches.</p>
<p>I hope that helps!</p>

**Comment by Molama** [comment]
<p>Thank you very much for that detailed and helpfull answer. </p>
<p>I did not know that there is a standard way to do such things, but that way is working really well for me.</p>
<p>Thank you and have a great day!</p>
